### PR TITLE
Upgrade MKL-DNN to version 0.17

### PR DESCRIPTION
**Pull request by gujinghui**
Upgrade MKL-DNN to version 0.17

### CODE REVIEW

**Comment by gujinghui** [comment]
@yinghai @ezyang 
pls help review.
looks like previously I made it too complicated. :(

@jgong5 @Jianhui-Li

**Comment by gujinghui** [comment]
rebased to fix build issue.

File: caffe2/ideep/operators/utility_ops.cc [hunk 1]
```diff
@@ -1,5 +1,3 @@
-#include "caffe2/operators/utility_ops.h"
-#include "caffe2/core/operator.h"
```
Comment (ezyang) [inline caffe2/ideep/operators/utility_ops.cc]: I don't see why this change should go in this patch

**Comment by ezyang** [comment]
Hi @gujinghui could you please tell us how you tested this patch

File: caffe2/ideep/operators/utility_ops.cc (hunk 1 above)
Comment (gujinghui) [reply caffe2/ideep/operators/utility_ops.cc]: just remove useless code.
if not good, I'll delete this change.

**Comment by gujinghui** [comment]
> Hi @gujinghui could you please tell us how you tested this patch

Not full understand your question.
Test it on local environment, and cannot repro previous issues now.

The issues described in #15345 are caused by mkl-dnn/cmake/MKL.cmake with wrong configuration.
In this PR, we skiped this cmake with our customized configurations.

**Comment by ezyang** [comment]
I just want to confirm; on the system you tested on, you were able to reproduce the previous problems with the old patch, is that correct? Can you post the new build log you get with this patch applied? (Also, does cmake reconfigure correctly the second time you run it?)

**Comment by gujinghui** [comment]
@ezyang 
Just uploaded the local build log. Pls help review.

[build.log](https://github.com/pytorch/pytorch/files/2697538/build.log)

[build2.log](https://github.com/pytorch/pytorch/files/2697542/build2.log)

**Comment by yinghai** [comment]
I think @ezyang would like to confirm:
1. Without this PR, could you reproduce the error?
2. With this PR, could you show that error is gone?

**Comment by gujinghui** [comment]
@yinghai 
clear to me now. Thanks.

No. I cannot repro the issues locally at all.
Otherwise, the previous PR should not be issued. :)
But according to the failed scenario, the problems should be fixed by this PR.

**Comment by ezyang** [comment]
```
-- Found MKLDNN: /data/users/ezyang/pytorch-repro-env/lib/libmkl_intel_lp64.so;/data/users/ezyang/pytorch-repro-env/lib/libm$
l_gnu_thread.so;/data/users/ezyang/pytorch-repro-env/lib/libmkl_core.so;-fopenmp;/usr/lib64/libpthread.so;/usr/lib64/libm.so$
/usr/lib64/libdl.so                               
-- Found OpenMP_C: -fopenmp (found version "3.1")   
-- Found OpenMP_CXX: -fopenmp (found version "3.1")
```

on a subsequent run, no MKL DNN configure lines occur.

**Comment by ezyang** [comment]
@gujinghui So, it appears that the problematic systems behave identically with this PR. However, I do not understand the cmake output. It says that it found MKLDNN, but all of the listed libraries afterward look like MKL, not MKLDNN. Can you explain this?

File: .jenkins/caffe2/build.sh [hunk 2]
```diff
@@ -132,6 +132,7 @@ CMAKE_ARGS+=("-DCMAKE_INSTALL_PREFIX=${INSTALL_PREFIX}")
 
 if [[ $BUILD_ENVIRONMENT == *mkl* ]]; then
   CMAKE_ARGS+=("-DBLAS=MKL")
+  CMAKE_ARGS+=("-DUSE_MKLDNN=ON")
```
Comment (ezyang) [inline .jenkins/caffe2/build.sh]: So... I don't think this should be necessary, because we should build with MKLDNN if we can find it. It looks like the problem is that setup.py and cmake have diverged in their defaults. In setup.py I see:

```

# ppc64le and aarch64 do not support MKLDNN
if IS_PPC or IS_ARM:
    USE_MKLDNN = check_env_flag('USE_MKLDNN', 'OFF')
else:
    USE_MKLDNN = check_env_flag('USE_MKLDNN', 'ON')
```

But in CMakeLists.txt I see

```
option(USE_MKLDNN "Use MKLDNN" OFF)
```

these should be made to agree.

**Comment by ezyang** [comment]
https://github.com/pytorch/pytorch/pull/15387/files#diff-d25356f3acd786db9e5a28c08ced1235R5

This comment is out of date and should be deleted.

File: cmake/Modules/FindMKLDNN.cmake [hunk 3]
```diff
@@ -41,28 +37,15 @@ ENDIF(NOT IDEEP_INCLUDE_DIR OR NOT MKLDNN_INCLUDE_DIR)
 LIST(APPEND MKLDNN_INCLUDE_DIR ${IDEEP_INCLUDE_DIR})
 
 IF(MKL_FOUND)
+  # Configure MKL-DNN
+  ADD_DEFINITIONS(-DUSE_MKL)
```
Comment (ezyang) [inline cmake/Modules/FindMKLDNN.cmake]: This line seems deeply questionable, for a number of reasons:

1. Is this for MKL-DNN? `USE_MKL` isn't even used by any source files in MKL-DNN. As far as I can tell, it is only used in `os_blas.h`, and that header is only included by headers (and no source files).
2. Is this for Caffe2? There aren't any uses of `USE_MKL` in Caffe2 either; we refer to it as `CAFFE2_USE_MKL`
3. Is it for fbgemm??? That seems super weird.

File: .jenkins/caffe2/build.sh (hunk 2 above)
Comment (gujinghui) [reply .jenkins/caffe2/build.sh]: @yinghai @ezyang 
I'm not sure with this.
If it's OK to build MKLDNN by default for caffe2 as well, I'll change it as above.

File: cmake/Modules/FindMKLDNN.cmake (hunk 3 above)
Comment (gujinghui) [reply cmake/Modules/FindMKLDNN.cmake]: @ezyang 
Yes. USE_MKL is for MKL-DNN build. The flag is used to enable 

> https://github.com/pytorch/pytorch/pull/15387/files#diff-d25356f3acd786db9e5a28c08ced1235R5
> 
> This comment is out of date and should be deleted.

Yes. I'll remove this comments line soon.

Comment (gujinghui) [reply cmake/Modules/FindMKLDNN.cmake]: > https://github.com/pytorch/pytorch/pull/15387/files#diff-d25356f3acd786db9e5a28c08ced1235R5
> 
> This comment is out of date and should be deleted.

Yes. I'll remove this comments line soon.

**Comment by gujinghui** [comment]
> https://github.com/pytorch/pytorch/pull/15387/files#diff-d25356f3acd786db9e5a28c08ced1235R5
> 
> This comment is out of date and should be deleted.

Yes. I'll remove this comments line soon.

**Comment by gujinghui** [comment]
> @gujinghui So, it appears that the problematic systems behave identically with this PR. However, I do not understand the cmake output. It says that it found MKLDNN, but all of the listed libraries afterward look like MKL, not MKLDNN. Can you explain this?

@ezyang 
That's weird. The error lines metioned in #15347 and #15345 have been removed in this PR.
Why the issue is still there? Could you send the log to me?
BTW, the warning message about not find MKL should be gone, right?

Yes. find_package_handle_standard_args in FindMKLDNN is used only to detect the MKLDNN dependencies, instead of MKL-DNN lib itself.
Because libmkldnn is built with caffe2 at the same time. There's no chance to find libmkldnn in this moment.

File: cmake/Modules/FindMKLDNN.cmake (hunk 3 above)
Comment (gujinghui) [reply cmake/Modules/FindMKLDNN.cmake]: @ezyang 
The USE_MKL is set for MKL-DNN build only.
It's related to the detail configuration of MKL-DNN.
Please refer to below comments copied from MKL-DNN src.
Currently, with better perf for caffe2, we set the config as, USE_MKL=ON && USE_CBLAS=OFF

 *  USE_MKL  USE_CBLAS effect
 *  -------  --------- ------
 *  yes      yes       normal compile: jit *may* be preferred over Intel(R) MKL CBLAS
 *  yes      no        jit calls OK; assert if cblas is ever called
 *  no       yes       system-dependent CBLAS
 *  no       no        gemm convolution (or other blas) N/A; create stubs

**Comment by gujinghui** [comment]
@ezyang 
The issue is very strange. And very confused because I still cannot repro it locally.
Did you have FB messager account? We can talk about it more detail on message.

**Comment by gujinghui** [comment]
@yinghai @ezyang 
I'm trying to static build MKL-DNN for caffe2 in #15504.
Suppose to fix the issues discussed in this PR as well.
If worked, pls close this PR and just merge #15504.

**Comment by syed-ahmed** [comment]
@ezyang @gujinghui Please note that upgrading to 0.17 won't solve the problem of dividing by zero (https://github.com/pytorch/pytorch/issues/15490), until this change is included in ideep/mkl-dnn:
https://github.com/herumi/xbyak/commit/7eb627509f4be89baa3a25b87dcfca1320273849#diff-0ec4be64047770b85c18baa80b10da90R142

**Comment by gujinghui** [comment]
#15504 is merged. This PR is not needed.
